<a href="https://colab.research.google.com/github/Mohan75489/Mohan75489/blob/main/Linear_Regression_Gradient_Descent.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Predicting a dimension model
#2 dependent & 3 independent variables for a data point
'''DataPoint1: 73,62,44 gives 56,70
DataPoint1: 90,84,60 gives 82,104
DataPoint1: 71,155,58 gives 117,143
DataPoint1: 109,40,32 gives 20,42
DataPoint1: 69,99,74 gives 106,123'''

'DataPoint1: 73,62,44 gives 56,70\nDataPoint1: 90,84,60 gives 82,104\nDataPoint1: 71,155,58 gives 117,143\nDataPoint1: 109,40,32 gives 20,42\nDataPoint1: 69,99,74 gives 106,123'

In [ ]:
#consider it as linear regression problem
#outputs (dependent variable) depends on each input(independent variables) with certain wait and a bias
'''Dim1=w11*dim1+w12*dim2+w13*dim3+b1
Dim2=w21*dim1+w22*dim2+w23*dim3+b2'''

'Dim1=w11*dim1+w12*dim2+w13*dim3\nDim2=w21*dim1+w22*dim2+w23*dim3'

In [4]:
#importing both numpy and PyTorch librarires
import torch
import numpy as np

Keeping one data point for checking model predictions.
Remaining 4 data points for training the model.

In [6]:
#taking inputs as numpy arrays and conversion to torch tensors
inputs = np.array([[73,62,44],[90,84,60],[71,155,58],[109,40,32]],dtype='float32')
targets = np.array([[56,70],[82,104],[117,143],[20,42]],dtype='float32')
inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)
print(inputs)
print(targets)

tensor([[ 73.,  62.,  44.],
        [ 90.,  84.,  60.],
        [ 71., 155.,  58.],
        [109.,  40.,  32.]])
tensor([[ 56.,  70.],
        [ 82., 104.],
        [117., 143.],
        [ 20.,  42.]])


In [7]:
#creating random weights and biases
w = torch.randn(3,2, requires_grad=True) #beacause 3 rows and 2 columns
b = torch.randn(2, requires_grad=True) #because 2 rows and 1 column
print(w)
print(b)

tensor([[ 0.5103,  0.7170],
        [-0.3340,  2.1479],
        [-0.6228,  0.1922]], requires_grad=True)
tensor([1.7046, 0.6569], requires_grad=True)


In [8]:
#creating a function to take inputs and return outputs
def model(x):
    return x @ w +b

In [9]:
#predicting first time (obviously too much far from actual values)
predict = model(inputs)
print(predict)

tensor([[ -9.1587, 194.6206],
        [-17.7977, 257.1374],
        [-49.9631, 395.6314],
        [ 24.0339, 170.8721]], grad_fn=<AddBackward0>)


In [10]:
#error or loss function (using mean sqaure error)
def mse(t1,t2):
      diff=t1-t2
      return torch.sum(diff*diff)/diff.numel()

In [11]:
#calculating loss using mse function created
loss = mse(predict,targets)
print(loss)

tensor(20188.7695, grad_fn=<DivBackward0>)


In [12]:
#loss is near to zero means model is predicting well (current model is too far from good)
#compute gradients, it's slope of loss function with respect to it's dependents (weights and biases).
loss.backward()
print(w)
print(w.grad)

tensor([[ 0.5103,  0.7170],
        [-0.3340,  2.1479],
        [-0.6228,  0.1922]], requires_grad=True)
tensor([[-6288.2676, 13715.8887],
        [-9535.1924, 16225.6914],
        [-4602.4053,  8362.0186]])


In [13]:
#adjusting weights and not calculating grads again (to increase efficiency)
with torch.no_grad():
    w -= w.grad * 1e-5
    b -= b.grad * 1e-5
    w.grad.zero_()
    b.grad.zero_()

In [14]:
#new weights and biases
print(w)
print(b)

tensor([[ 0.5732,  0.5798],
        [-0.2387,  1.9856],
        [-0.5768,  0.1085]], requires_grad=True)
tensor([1.7055, 0.6553], requires_grad=True)


In [15]:
#one more iteration of loss calculation
predict = model(inputs)
loss = mse(predict, targets)
print(loss)

tensor(14051.3076, grad_fn=<DivBackward0>)


We can observe loss is decreased drastically from 20188 to 14051.

In [18]:
#runnning in a for loop to iterate loss function for 100 times (epochs)
for i in range(100):
    predict = model(inputs)
    loss = mse(predict, targets)
    loss.backward()
    with torch.no_grad():
        w -= w.grad * 1e-5
        b -= b.grad * 1e-5
        w.grad.zero_()
        b.grad.zero_()

In [19]:
#check the new loss after epochs
predict = model(inputs)
loss = mse(predict, targets)
print(loss)

tensor(202.1741, grad_fn=<DivBackward0>)


In [20]:
#we can check accuracy by comparing targets and predictions
print(targets)
print(predict)

tensor([[ 56.,  70.],
        [ 82., 104.],
        [117., 143.],
        [ 20.,  42.]])
tensor([[ 51.7374,  64.5638],
        [ 67.1953,  85.9993],
        [113.7678, 155.9432],
        [ 48.1984,  49.3038]], grad_fn=<AddBackward0>)


In [22]:
#seems good for some data points, probably it need 100 more epochs
for i in range(100):
    predict = model(inputs)
    loss = mse(predict, targets)
    loss.backward()
    with torch.no_grad():
        w -= w.grad * 1e-5
        b -= b.grad * 1e-5
        w.grad.zero_()
        b.grad.zero_()

In [24]:
predict = model(inputs)
loss = mse(predict,targets)
print(loss)
print(targets)
print(predict)

tensor(135.0044, grad_fn=<DivBackward0>)
tensor([[ 56.,  70.],
        [ 82., 104.],
        [117., 143.],
        [ 20.,  42.]])
tensor([[ 49.7195,  65.9598],
        [ 66.0113,  87.6554],
        [123.0549, 151.9084],
        [ 36.6851,  52.3474]], grad_fn=<AddBackward0>)


In [25]:
#we can refain it more, this time for 200 epochs with same learning rate
for i in range(200):
    predict = model(inputs)
    loss = mse(predict, targets)
    loss.backward()
    with torch.no_grad():
        w -= w.grad * 1e-5
        b -= b.grad * 1e-5
        w.grad.zero_()
        b.grad.zero_()

In [26]:
predict = model(inputs)
loss = mse(predict,targets)
print(loss)
print(targets)
print(predict)

tensor(103.7898, grad_fn=<DivBackward0>)
tensor([[ 56.,  70.],
        [ 82., 104.],
        [117., 143.],
        [ 20.,  42.]])
tensor([[ 49.9876,  67.0929],
        [ 67.2147,  89.4720],
        [124.8090, 149.9828],
        [ 32.0629,  52.0397]], grad_fn=<AddBackward0>)


In [27]:
for i in range(1000):
    predict = model(inputs)
    loss = mse(predict, targets)
    loss.backward()
    with torch.no_grad():
        w -= w.grad * 1e-5
        b -= b.grad * 1e-5
        w.grad.zero_()
        b.grad.zero_()

In [28]:
predict = model(inputs)
loss = mse(predict,targets)
print(loss)
print(targets)
print(predict)

tensor(32.0964, grad_fn=<DivBackward0>)
tensor([[ 56.,  70.],
        [ 82., 104.],
        [117., 143.],
        [ 20.,  42.]])
tensor([[ 53.4409,  70.1589],
        [ 73.5686,  95.0636],
        [121.2771, 146.7034],
        [ 26.3148,  47.2312]], grad_fn=<AddBackward0>)


In [29]:
for i in range(1000):
    predict = model(inputs)
    loss = mse(predict, targets)
    loss.backward()
    with torch.no_grad():
        w -= w.grad * 1e-5
        b -= b.grad * 1e-5
        w.grad.zero_()
        b.grad.zero_()

In [30]:
predict = model(inputs)
loss = mse(predict,targets)
print(loss)
print(targets)
print(predict)

tensor(11.3510, grad_fn=<DivBackward0>)
tensor([[ 56.,  70.],
        [ 82., 104.],
        [117., 143.],
        [ 20.,  42.]])
tensor([[ 55.3207,  71.8001],
        [ 77.0057,  98.0650],
        [119.2917, 144.9698],
        [ 23.3146,  44.6115]], grad_fn=<AddBackward0>)


In [34]:
#now I am satisfied with loss @ 11, now time to pass 5th data point
inputs=torch.tensor([68,99,74.])
predict = model(inputs)
loss = mse(predict,targets)
print(predict)

tensor([103.2584, 121.1408], grad_fn=<AddBackward0>)


Prediction is very close to target. Great!!!

In [35]:
#Linear regression using PyTorch built-ins
import torch.nn as nn #PyTorch utility classes for building neural networks

In [36]:
#taking a bigger data set using numpy and converting to tensors of PyTorch
# Input (dim1, dim2, dim3)
inputs = np.array([[73, 67, 43], 
                   [91, 88, 64], 
                   [87, 134, 58], 
                   [102, 43, 37], 
                   [69, 96, 70], 
                   [74, 66, 43], 
                   [91, 87, 65], 
                   [88, 134, 59], 
                   [101, 44, 37], 
                   [68, 96, 71], 
                   [73, 66, 44], 
                   [92, 87, 64], 
                   [87, 135, 57], 
                   [103, 43, 36], 
                   [68, 97, 70]], 
                  dtype='float32')

# Targets (Dim1, Dim2)
targets = np.array([[56, 70], 
                    [81, 101], 
                    [119, 133], 
                    [22, 37], 
                    [103, 119],
                    [57, 69], 
                    [80, 102], 
                    [118, 132], 
                    [21, 38], 
                    [104, 118], 
                    [57, 69], 
                    [82, 100], 
                    [118, 134], 
                    [20, 38], 
                    [102, 120]], 
                   dtype='float32')

inputs = torch.from_numpy(inputs)
targets = torch.from_numpy(targets)

In [37]:
#to define the dataset
from torch.utils.data import TensorDataset
train_ds = TensorDataset(inputs,targets)

In [38]:
#to take random batch of given inputs to train the model
from torch.utils.data import DataLoader
bs = 5
train = DataLoader(train_ds,bs,shuffle=True) #enble shuffling to train with random data

In [41]:
#defining a model using nn and wirghts & biases intialization
model = nn.Linear(3,2)   # 3 inputs and 2 outputs
#we can see intialized weights and biases
print(model.weight)
print(model.bias)
#.parameter is also used to call weights and bias from model
list(model.parameters())

Parameter containing:
tensor([[ 0.5067,  0.1565,  0.5292],
        [-0.2564, -0.0915,  0.5316]], requires_grad=True)
Parameter containing:
tensor([-0.1942, -0.2274], requires_grad=True)


[Parameter containing:
 tensor([[ 0.5067,  0.1565,  0.5292],
         [-0.2564, -0.0915,  0.5316]], requires_grad=True),
 Parameter containing:
 tensor([-0.1942, -0.2274], requires_grad=True)]

In [44]:
# intial prediction
preds = model(inputs)
preds

tensor([[ 70.0382,  -2.2142],
        [ 93.5593,   2.4133],
        [ 95.5580,  -3.9596],
        [ 77.7999, -10.6431],
        [ 86.8398,  10.5115],
        [ 70.3883,  -2.3791],
        [ 93.9320,   3.0365],
        [ 96.5939,  -3.6843],
        [ 77.4498, -10.4782],
        [ 86.8623,  11.2994],
        [ 70.4109,  -1.5911],
        [ 93.9095,   2.2485],
        [ 95.1853,  -4.5827],
        [ 77.7774, -11.4311],
        [ 86.4897,  10.6763]], grad_fn=<AddmmBackward0>)

Not Great. Obviously!

In [45]:
#importing nn.fuctional to get fuctions
import torch.nn.functional as F
loss = F.mse_loss #defining loss function as mean square error
losses = loss(model(inputs),targets) #ccalculating loss
print(losses)

tensor(5197.2202, grad_fn=<MseLossBackward0>)


In [46]:
#using PyTorch optim defining optimizer model
opt = torch.optim.SGD(model.parameters(), lr = 1e-6)   #SGD = Stochastic gradient descent

In [52]:
#training the model and printing every 10th epoch loss
#it uses data loaded previous level and all funxtions created above
#model, loss, opt
def val(num_epochs, model, loss, opt, train):
    #Repeat for given number of epochs
    for epoch in range(num_epochs):
      for x,y in train:
          # 1. Generating predictions
          pred = model(x)
          # 2. Calculating loss
          losses = loss(pred,y)
          # 3. Computing gradients
          losses.backward()
          # 4. Upating parameters (weights and biases) using SGD
          opt.step()
          # 5. Resetting the gradients to zero
          opt.zero_grad()
      #printing the losses for every 10 epoch
      if(epoch+1) % 10 ==0:
        print('Epoch [{}/{}], Loss: {}'.format(epoch+1,num_epochs,losses.item()))

In [53]:
val(1000,model,loss,opt,train)

Epoch [10/1000], Loss: 18.300861358642578
Epoch [20/1000], Loss: 5.664689064025879
Epoch [30/1000], Loss: 9.919859886169434
Epoch [40/1000], Loss: 4.524773597717285
Epoch [50/1000], Loss: 3.6833317279815674
Epoch [60/1000], Loss: 9.3099365234375
Epoch [70/1000], Loss: 11.728381156921387
Epoch [80/1000], Loss: 10.106321334838867
Epoch [90/1000], Loss: 14.023469924926758
Epoch [100/1000], Loss: 13.7827730178833
Epoch [110/1000], Loss: 13.824953079223633
Epoch [120/1000], Loss: 8.832287788391113
Epoch [130/1000], Loss: 11.218454360961914
Epoch [140/1000], Loss: 6.14157772064209
Epoch [150/1000], Loss: 12.180593490600586
Epoch [160/1000], Loss: 4.422818660736084
Epoch [170/1000], Loss: 6.061863899230957
Epoch [180/1000], Loss: 5.067444801330566
Epoch [190/1000], Loss: 4.9310712814331055
Epoch [200/1000], Loss: 7.262157440185547
Epoch [210/1000], Loss: 9.708313941955566
Epoch [220/1000], Loss: 6.693589687347412
Epoch [230/1000], Loss: 8.6727933883667
Epoch [240/1000], Loss: 7.53333568572998

In [54]:
#comparing predictions and targtets
preds = model(inputs)
print(preds)
print(targets)

tensor([[ 57.1583,  70.4669],
        [ 81.8314, 100.6035],
        [118.3437, 132.5222],
        [ 22.1629,  38.3996],
        [100.6844, 118.1936],
        [ 55.9398,  69.4085],
        [ 81.6338, 100.7017],
        [118.6194, 133.1329],
        [ 23.3814,  39.4579],
        [101.7053, 119.3501],
        [ 56.9607,  70.5651],
        [ 80.6129,  99.5451],
        [118.5413, 132.4240],
        [ 21.1421,  37.2430],
        [101.9029, 119.2519]], grad_fn=<AddmmBackward0>)
tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 57.,  69.],
        [ 80., 102.],
        [118., 132.],
        [ 21.,  38.],
        [104., 118.],
        [ 57.,  69.],
        [ 82., 100.],
        [118., 134.],
        [ 20.,  38.],
        [102., 120.]])


In [56]:
#testing with any 1 data point randomly and new one
model(torch.tensor([88,134,59.]))

tensor([118.6194, 133.1329], grad_fn=<AddBackward0>)

In [57]:
model(torch.tensor([75,63,44.]))

tensor([53.6779, 67.6630], grad_fn=<AddBackward0>)

Great!!!. It's predicting too good.